### ISSUE TITLE
modify `ls.available()` to include each snippets origin file path.

### ISSUE THREAD

**Issue by molleweide** [L3MON4D3/LuaSnip#717]
### Discussed in https://github.com/L3MON4D3/LuaSnip/discussions/716

<div type='discussions-op-text'>

<sup>Originally posted by **molleweide** January  6, 2023</sup>
## Proposed solution

After digging in the source a little bit I am thinking that the appropriate way to go about this would be 

1. Attach the `origin_file_path` to each snippet when loading files with `load_files(ft, files)` from `luasnip/loaders/from_lua.lua`. 
2. Update the `luasnip/init.lua` func `get_context()` so that it also grabs the path of each snip.
3. Modify the available function to mean "available in a broader sense".
    ```lua
    local function available(ft:<string|table>)        
    -- ft could be either a string or a table.
    -- 1. if ft == nil, then do the exact same as current `available()` and deprecate the available function.
    -- 2. if ft == string, retrieve snips for this specific file type.
    -- 3. if ft == table, then return all snippets for the containing filetypes.
    ```
4. optionally, also do this same procedure for `snipmate`, and `vscode` as well.
     
With these three additions it would be possible to extend the existing `telescope-luasnip` extension repo to become a CRUD manager. I also have some more ideas for the picker that would make life easier, eg. add snippet to new filetype and then create all of the boilerplate in your default luasnippets path for a new filetype.

If @L3MON4D3 if you agree with my proposed solution I will make a PR for this. This should NOT introduce any breaking changes.</div>

**Comment by L3MON4D3** [comment]
> (I tried the "transfer" feature and assumed that the existing discussion would be removed...)

No worries, I'll leave a comment to redirect here :D

I think that's a pretty cool usecase, we already have open-by-filetype, open-by-snippet is a slick addition.

But! I would propose to implement this by adding `source` (or `metadata`, for arbritary data, and add a config-option for what is put in there) as a key to `add_snippets`, this would make it more versatile (can also be done if a snippets is added via `add_snippets`).
Not sure what you are thinking of with attaching the source to the snippet, but I'd rather not include metadata inside the snippet itself. I think a table mapping `snippet-id -> source/metadata`+ a function to retrieve that should do it.

Creating the boilerplate could be done by [passing `extend` in `edit_snippet_files`'s `opts`](https://github.com/L3MON4D3/Dotfiles/blob/b974253b38a53768f5a89e965b4f3b6744f65e5c/.config/nvim/lua/plugins/luasnip/ft_edit.lua#L9-L28), which is pretty elegant imo, but similar functionality can be added to `telescope-luasnip`, if that  fits your workflow better :D

**Comment by molleweide** [comment]
Thanks for taking my idea seriously ;)

> I think that's a pretty cool usecase, we already have open-by-filetype, open-by-snippet is a slick addition.

Yeah right, because it would reduce number of strokes to get to a snip by a lot.

> But! I would propose to implement this by adding `source` (or `metadata`, for arbritary data, and add a config-option for what is put in there) as a key to `add_snippets`, this would make it more versatile (can also be done if a snippets is added via `add_snippets`).

Aaah yes, I see what you mean. 

> Not sure what you are thinking of with attaching the source to the snippet, but I'd rather not include metadata inside the snippet itself. I think a table mapping `snippet-id -> source/metadata`+ a function to retrieve that should do it.

Well for my own picker I just made a custom loader and did `snippet["origin_file_path"] = ..." but I understand that using this retrieval method with an id seems more grown up haha.

> Creating the boilerplate could be done by [passing `extend` in `edit_snippet_files`'s `opts`](https://github.com/L3MON4D3/Dotfiles/blob/b974253b38a53768f5a89e965b4f3b6744f65e5c/.config/nvim/lua/plugins/luasnip/ft_edit.lua#L9-L28), which is pretty elegant imo, but similar functionality can be added to `telescope-luasnip`, if that fits your workflow better :D

I have never tried this you describe here. I have to do some playing around and then I will get back here with an updated solution.

**Comment by molleweide** [comment]
## progress so far

https://github.com/molleweide/LuaSnip/commits/molleweide

**Comment by molleweide** [comment]
### treesitter idea 
     1. require ts 
     2. local snip_tree = query for snippet containing name or trig string
     3. depending on which mapping was used filter out required subcomponent.
     4. move cursor to snippet subcomponent, eg. node x etc.

**Comment by leiserfg** [comment]
Considering the amount of data, should the "storing metadata" part be opt-in?

**Comment by molleweide** [comment]
> Considering the amount of data, should the "storing metadata" part be opt-in?  

Good idea, and maybe one could also configure which specific dirs to add source maps for.

**Comment by L3MON4D3** [comment]
> The solution should ofc work regardless of telescope . If you have the completion menu open or even if you are within an active snippet, we could have an insert mode mapping that (rolls back the snippet and) exits to normal mode and opens up the current snippet in a split or something.

Right, `delete faulty snippet -> edit source of snippet` sounds like a good workflow :+1:



> ### crud (brainstorming)
> When the `jump-to-snip` feature is working, what do you think about having something like `utils/crud.lua OR manage.lua` where we could put all functions related to managing single snippets. Eg.:
> 
> 1. jump to snip
> 2. remove snip
> 3. update -> eg. you make a bind for updateing the trig or descr and then move cursor and enter insert with treesitter etc.
>    Basically make a comprehensive manager function that allows you to jump specifically to any node within a snippet instantly and then take it from there with whatever user `opts` or configs in order to reduce keystrokes as much as possible.
> 
> In telescope or insert mode we could have:
> 
> * `<C-u><C-t>` jump to trig string
>   -`<C-u><C-i>` jump to first insert node
>   etc. so that we achieve a super fine grained control of jumping to snip.

I don't think it's worth it to reduce keystrokes beyond moving the cursor onto the line where the snippet is defined. Stuff like remove/updating some part of a snippet seems too specific to me.



> ### treesitter idea / crud api
> ```
>  1. require ts 
>  2. local snip_tree = query for snippet containing name or trig string
>  3. depending on which mapping was used filter out required subcomponent.
>  4. create some api funcs
>      - move cursor to snip
>      - add new snippet to same snip file as selected and expand template snip.
>      - remove snip 
>      - move snip from snip file A to B (not super important lol)
>      - move cursor to snip node N
>      - if has `fmt()` move cursor to format string
> ```
> 
> Which would be the most useful ones? Do you have any additional ideas you'd like?

Same as above, anything beyond finding where the snippet is defined seems too granular. But, if you want to implement this, don't let my opinion stop you :D I'm not convinced it's useful, but maybe my workflow is just different from yours

> Considering the amount of data, should the "storing metadata" part be opt-in?

Right, yes, it probably should. I'm thinking of adding a config-option `loader_snip_metadata` (name WIP xD) which takes a table like `{lua = ..., vscode = ..., snipmate = ...}`, and `...` are functions which get as much data as a loader can provide (only the file I think), and generate the metatdata that should be attached to its snippets. By default `function() return nil end` such that nothing is stored, we could, as usual :D, provide some sensible defaults in `luasnip.extras`. This would also cover excluding some dirs.

How to store it, yeah, that's a question.. Most straightforward would be snip_id -> metadata, but that might be a bit too wasteful. Perhaps snip_id -> metadata_id -> metadata?

**Comment by molleweide** [comment]
> I'm not convinced it's useful,

Lol, me neither. I consider this RnD. Now that I got the jump to snip functionality to work I feel pretty much satisfied because it is very fast (and fast enough..). However, I might play around with my picker and see what happens. Just like say, it is stupid to over optimise so we'll see what happens maybe some patterns arise that become useful. Also, I'm less experienced as a coder/luasnipper so I really need to consult with you before I spend time on something really stupid.

Anyways, I'm thinking that now I'll 
1. add a config option for my solution that defaults as `false`, 
2. add this option to the docs, 
3. and then I'll submit the PR so that we can have this solution working for now until we might come up with something smarter in the future.

Does this sound good?

**Comment by L3MON4D3** [comment]
> Lol, me neither. I consider this RnD

Ah, alright xD Go ahead then :D

> Does this sound good?

Yeup :+1:
I'll wait for your PR and then take a closer look at the implementation

One more idea: make it possible to attach more metadata to the snippets. But for now, it should be enough to have one metatdata-entry per `add_snippets`-call, and then point all snippets to that

**Comment by molleweide** [comment]
PR submitted.

> make it possible to attach more metadata to the snippets.

I believe that my implementation makes this possible. I created a table `{ source = path}` for each add_snippets call and then point each snip to this table.
